BCH Integration -- Updated Admin-UI and TestSetup for both BTC/BCH

cnohall opened the pull request:
# BCH Integration - Admin UI and TestSetup
This PR is related to issue #105. This PR introduces a new _Currencies_ section that consists of two checkboxes. One for BTC and one for BCH. The _Save_ button in the Settings section and the _Test Setup_ in the Currencies section save the values from both sections, with the exception that pressing _Test Setup_ also runs `TestSetup()` for the active cryptos. Which values are saved or not is easy to adapt, if need be. 

### Current design
![image](https://user-images.githubusercontent.com/36883813/117113218-21132f00-adc5-11eb-9cda-246e5568af5d.png)

### Code changes
`blockonomics.php` has been quite heavily modified, to accommodate the new section and test setup for both cryptos. The code was written according to the pattern set in the WooCommerce-plugin.

`validation.php` only has a few new comments and a changed function name. 

The module has been tested and the correct error messages are shown at the right time. It also correctly updates the callback for both BTC and BCH. The checkout still works. However, since we don't have a selection screen at the moment the checkout will always create an order in BTC.

The module is in line with PrestaShop's standards.

DarrenWestwood reviewed (commented):
The pull request has good reuse of code from the Woocommerce plugin.

cnohall commented on blockonomics.php:
```diff
@@ -67,6 +67,17 @@ public function __construct()
         $BLOCKONOMICS_SET_CALLBACK_URL =
             $BLOCKONOMICS_BASE_URL . '/api/update_callback';
 
+        $BCH_BLOCKONOMICS_BASE_URL = 'https://bch.blockonomics.co';
```
Done

cnohall commented on blockonomics.php:
```diff
@@ -288,18 +334,80 @@ public function doCurlCall($url, $post_content = '')
 
+    public function checkCallbackUrlsOrSetOne($crypto, $response)
+    {
+        //check the current callback and detect any potential errors
```
Done.

cnohall commented on blockonomics.php:
```diff
@@ -512,6 +624,53 @@ class="process-icon-refresh"></a>
+                    'type' => 'checkbox',
+                    'label'     => $this->l('Bitcoin (BTC)'),
+                    'desc'      => $desc .
+                    '<a href="https://blockonomics.co/merchants">https://blockonomics.co/merchants</a>',
```
Done

cnohall commented on blockonomics.php:
```diff
@@ -406,38 +530,37 @@ public function getContent()
+            $error_strings = $this->testSetup();
+            $api_key = Configuration::get('BLOCKONOMICS_API_KEY');
+            //if there's no API key, give error immediately
+            if (!$api_key) {
```
True, I've moved the `api_key`-check to be before testSetup().

DarrenWestwood commented on blockonomics.php:
```diff
@@ -288,18 +334,80 @@ public function doCurlCall($url, $post_content = '')
+        $error_str = $this->checkCallbackUrlsOrSetOne($crypto, $response);
+        if (!$error_str) {
+            //Everything OK ! Test address generation
+            $response = $this->getNewAddress($crypto);
```
test_mode should be used to prevent Test Setup generating a new address with every click.
`$response = $this->getNewAddress($crypto, true);`

cnohall replied:
Done, I've set it to true.

shivaenigma commented:
Lets remove [this message](https://prnt.sc/13zc1de) not necessary
>API Key is not provided to communicate with Blockonomics
change to 
Please specify an API Key

cnohall commented:
I've now removed [this message](https://prnt.sc/13zc1de) and updated

> API Key is not provided to communicate with Blockonomics 

to 

> Please specify an API Key

cnohall commented on blockonomics.php:
```diff
@@ -406,38 +528,37 @@ public function getContent()
+                $error_strings = $this->testSetup();
+                foreach ($error_strings as $crypto => $error_str) {
+                    if ($error_str) {
+                        $article_url = 'https://blockonomics.freshdesk.com/solution/articles/';
```
Done.

cnohall commented on blockonomics.php:
```diff
@@ -67,6 +67,17 @@ public function __construct()
             $BLOCKONOMICS_BASE_URL . '/api/update_callback';
 
+        $BLOCKONOMICS_BCH_BASE_URL = 'https://bch.blockonomics.co';
+        $BLOCKONOMICS_BCH_NEW_ADDRESS_URL =
```
True, this should probably be implemented in the WordPress plugin too. Done and pushed

cnohall commented on blockonomics.php:
```diff
@@ -247,16 +289,20 @@ public function getBTCPrice($id_currency)
+            $new_address_url = Configuration::get('BLOCKONOMICS_NEW_ADDRESS_URL');
+        } else {
+            $new_address_url = Configuration::get('BLOCKONOMICS_BCH_NEW_ADDRESS_URL');
+        }
```
Done

cnohall replied:
I have an idea of further simplification, I'll test it out

cnohall commented on blockonomics.php:
```diff
@@ -309,71 +417,85 @@ public function testSetup()
+            $set_callback_url = Configuration::get('BLOCKONOMICS_SET_CALLBACK_URL');
+        } else {
+            $set_callback_url = Configuration::get('BLOCKONOMICS_BCH_SET_CALLBACK_URL');
+        }
```
Done

cnohall commented on blockonomics.php:
```diff
@@ -309,71 +417,85 @@ public function testSetup()
+            $url = Configuration::get('BLOCKONOMICS_GET_CALLBACKS_URL');
+        } else {
+            $url = Configuration::get('BLOCKONOMICS_BCH_GET_CALLBACKS_URL');
+        }
```
Done

cnohall commented:
On @shivaenigma's initiative, I introduced `getServerAPIURL()`:
```
public function getServerAPIURL($crypto, $path) {
        $domain = ($crypto == 'btc') ? 'www' : 'bch';
        return "https://" . $domain . ".blockonomics.co" . $path;
} 
```
There are four different paths, `BLOCKONOMICS_PRICE_URL` `BLOCKONOMICS_NEW_ADDRESS_URL` `BLOCKONOMICS_GET_CALLBACKS_URL` `BLOCKONOMICS_SET_CALLBACKS_URL`. Each path is only used once. Instead of saving each path as a variable in the configuration, they're stored in their respective function. This saves us from defining variables, updating and deleting them. 

`$BLOCKONOMICS_WEBSOCKET_URL` still remains, and can be removed. However, since it's not connected to testSetup() or the Admin-ui, I suggest that make those changes in the next PR.

shivaenigma commented on blockonomics.php:
```diff
@@ -309,71 +353,78 @@ public function testSetup()
+
+    public function getCallbacks($crypto)
+    {
+        $path = '/api/address?&no_balance=true&only_xpub=true&get_callback=true';
```
Declaring path like this is ok. But we are uncessary creating a new variable which is a constant on each function call. This is why programming guides recommend using way to declare a *global/class scope const string* variable. This way memory for this variable is allocated only once

Also it helps to have all these const strings at one place. Similar to the translate string that we are declaring

cnohall replied:
True, I've added defined the strings as _global/class scope const string_. I also removed three lines of `Configuration::deleteByName` for variables that never were introduced.

As a side note, the WooCommerce plugin can be cleaned from code duplication in a similar way.
